lightning/src/offers/contacts.rs

```diff
@@ -0,0 +1,366 @@
+/// We verify that the offer uses this signing key, otherwise the BIP 353 address most likely
+/// doesn't belong to them.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct UnverifiedContactAddress {
+	address: ContactAddress,
+	expected_offer_signing_key: PublicKey,
+}
+
+// FIXME: this can be simply a function call?
+impl UnverifiedContactAddress {
+	/// Creates a new [`UnverifiedContactAddress`].
+	pub fn new(address: ContactAddress, expected_offer_signing_key: PublicKey) -> Self {
+		Self { address, expected_offer_signing_key }
+	}
+
+	/// Returns the contact address.
+	pub fn address(&self) -> &ContactAddress {
+		&self.address
+	}
+
+	/// Returns the expected offer signing key.
+	pub fn expected_offer_signing_key(&self) -> PublicKey {
+		self.expected_offer_signing_key
+	}
+
+	/// Verify that the offer obtained by resolving the BIP 353 address matches the
+	/// invoice_request commitment.
+	///
+	/// If this returns false, it means that either:
+	///  - the contact address doesn't belong to the node
+	///  - or they changed the signing key of the offer associated with their BIP 353 address
+	///
+	/// Since the second case should be very infrequent, it's more likely that the remote node
+	/// is malicious and we shouldn't store them in our contacts list.
+	pub fn verify(&self, offer: &Offer) -> bool {
+		// Check if the expected key matches the offer's issuer ID
+		if let Some(issuer_id) = offer.issuer_signing_pubkey() {
+			if issuer_id == self.expected_offer_signing_key {
+				return true;
+			}
+		}
+
+		// Check if the expected key matches any of the blinded path node IDs
+		for path in offer.paths() {
+			if let IntroductionNode::NodeId(node_id) = path.introduction_node() {
+				if *node_id == self.expected_offer_signing_key {
+					return true;
+				}
+			}
+		}
+
+		false
+	}
+}
+
+/// Contact secrets are used to mutually authenticate payments.
+///
+/// The first node to add the other to its contacts list will generate the `primary_secret` and
+/// send it when paying. If the second node adds the first node to its contacts list from the
+/// received payment, it will use the same `primary_secret` and both nodes are able to identify
+/// payments from each other.
+///
+/// But if the second node independently added the first node to its contacts list, it may have
+/// generated a different `primary_secret`. Each node has a different `primary_secret`, but they
+/// will store the other node's `primary_secret` in their `additional_remote_secrets`, which lets
+/// them correctly identify payments.
+///
+/// When sending a payment, we must always send the `primary_secret`.
+/// When receiving payments, we must check if the received contact_secret matches either the
+/// `primary_secret` or any of the `additional_remote_secrets`.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ContactSecrets {
+	primary_secret: [u8; 32],
+	additional_remote_secrets: Vec<[u8; 32]>,
```

Comment:
Should this be a single `Option`? I'm a bit unclear on what UX would result in there being more than one.